File: rules-interpreter/src/main/java/com/github/iamrenny/ruleflow/visitors/Visitor.java
package com.github.iamrenny.ruleflow.visitors;



import com.github.iamrenny.ruleflow.RuleFlowLanguageBaseVisitor;
import com.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import com.github.iamrenny.ruleflow.errors.PropertyNotFoundException;
import com.github.iamrenny.ruleflow.evaluators.*;
import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ParseTree;

import java.util.Map;
import java.util.Set;

public class Visitor extends RuleFlowLanguageBaseVisitor<Object> {
    private final Map<String, ?> data;
    private final Map<String, Set<String>> lists;
    private final Map<String, ?> root;


    public Visitor(Map<String, ?> data, Map<String, Set<String>> lists, Map<String, ?> root) {
        this.data = data;
        this.lists = lists != null ? lists : Map.of();
        this.root = root;
    }

    @Override
    public Object visit(ParseTree tree) {
        ParserRuleContext ctx = (ParserRuleContext) tree;

        ContextEvaluator<?> condition;

        if (ctx instanceof RuleFlowLanguageParser.ComparatorContext) {
            condition = new ComparatorContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.AggregationContext) {
            condition = new AggregationContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.MathMulContext) {
            condition = new MathMulContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.MathAddContext) {
            condition = new MathAddContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.ParenthesisContext) {
            condition = new ParenthesisContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.ValueContext) {
            condition = new ValueContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.PropertyContext) {
            condition = new PropertyContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.ValidPropertyContext) {
            condition = new ValidPropertyContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.DateDiffContext) {
            condition = new DateDiffContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.ListContext) {
            condition = new ListContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.UnaryContext) {
            condition = new UnaryContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.BinaryAndContext) {
            condition = new BinaryAndContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.BinaryOrContext) {
            condition = new BinaryOrContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.DayOfWeekContext) {
            condition = new DayOfWeekContextEvaluator();
        } else if (ctx instanceof RuleFlowLanguageParser.RegexlikeContext) {
            condition = new RegexContextEvaluator();
        } else {
            throw new IllegalArgumentException("Operation not supported: " + ctx.getClass());
        }

        try {
            return condition.evaluate(ctx, this);
        } catch (PropertyNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    public Map<String, ?> getData() {
        return data;
    }

    public Map<String, Set<String>> getLists() {
        return lists;
    }

    public Map<String, ?> getRoot() {
        return root;
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/fraud/rules/parser/visitors/Visitor.kt
package com.rappi.fraud.rules.parser.visitors

import com.rappi.fraud.analang.ANABaseVisitor
import com.rappi.fraud.analang.ANAParser
import com.rappi.fraud.rules.parser.evaluators.AggregationContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.BinaryAndContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.BinaryOrContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.ComparatorContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.ContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.DateDiffContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.ListContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.MathAddContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.MathMulContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.ParenthesisContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.PropertyContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.UnaryContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.ValidPropertyContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.ValueContextEvaluator
import com.rappi.fraud.rules.parser.evaluators.DayOfWeekContextEvaluator
import org.antlr.v4.runtime.ParserRuleContext
import org.antlr.v4.runtime.tree.ParseTree


class Visitor(
    val data: Map<String, *>,
    val lists:  Map<String, Set<String>> = mapOf(),
    val root: Map<String, *>
) : ANABaseVisitor<Any>() {
    override fun visit(tree: ParseTree?): Any {
        val ctx = tree as ParserRuleContext

        val condition = when (ctx) {
            is ANAParser.ComparatorContext -> ComparatorContextEvaluator()
            is ANAParser.AggregationContext -> AggregationContextEvaluator()
            is ANAParser.MathMulContext -> MathMulContextEvaluator()
            is ANAParser.MathAddContext -> MathAddContextEvaluator()
            is ANAParser.ParenthesisContext -> ParenthesisContextEvaluator()
            is ANAParser.ValueContext -> ValueContextEvaluator()
            is ANAParser.PropertyContext -> PropertyContextEvaluator()
            is ANAParser.ValidPropertyContext -> ValidPropertyContextEvaluator()
            is ANAParser.DateDiffContext -> DateDiffContextEvaluator()
            is ANAParser.ListContext -> ListContextEvaluator()
            is ANAParser.UnaryContext -> UnaryContextEvaluator()
            is ANAParser.BinaryAndContext -> BinaryAndContextEvaluator()
            is ANAParser.BinaryOrContext -> BinaryOrContextEvaluator()
            is ANAParser.DayOfWeekContext -> DayOfWeekContextEvaluator()
            else -> throw IllegalArgumentException("Operation not supported: ${ctx.javaClass}")
        } as ContextEvaluator<ParserRuleContext>

        return condition.evaluate(ctx, this)
    }
}